tensorflow/compiler/xla/service/pattern_matcher.h

```diff
@@ -2153,8 +2158,11 @@ class HloInstructionPattern {
     return AppendImpl(HloInstructionPatternComparisonDirectionImpl(direction));
   }
 
-  auto WithPredicate(HloPredicate fn) const {
-    return AppendImpl(HloInstructionPredicateImpl(std::move(fn)));
+  // Modifies the pattern to match only if fn(instruction, args...) returns
+  // true.
+  template <typename Fn, typename... Args>
+  auto WithPredicate(Fn fn, Args&... args) const {
```

Comment:
We can std::forward here but I don't think the class template parameter of HloInstructionPredicateImpl can become a forwarding reference.

Reply:
I'm not very familiar with perfect forwarding, but yeah, it seems difficult to use perfect forwarding with the `std::tuple` field of HloInstructionPredicateImpl. I'm not sure if this is possible or not.

I'm more convinced now that `WithPredicate` should not support arguments.

Reply:
Only function template parameters can become forwarding references.

Isn't FindF8SubgraphRecursive too complex for a lambda?

Reply:
> Only function template parameters can become forwarding references.

Ah, I didn't realize.

> Isn't FindF8SubgraphRecursive too complex for a lambda?

I mean to wrap it in a lambda, e.g. instead of

```
m::Op().WithPredicate(FindF8SubgraphRecursive, a, a_scale,
  a_unary_ops, a_mult_scale),
```

You would do something like

```
m::Op().WithPredicate([&](const HloInstruction& instr) {
  return FindF8SubgraphRecursive(instr, a, a_scale, a_unary_ops, a_mult_scale);
  }),
```

Reply:
That could be a good solution.